### PR TITLE
on Windows first search for SDL2 libraries, then download from libsdl.org

### PR DESCRIPTION
**Pull request by rfomin**
SDL_Image is always downloaded from libsdl.org on Windows. That miniz is not that mini, it feels weird to add so much irrelevant code to Doom port git.

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
> SDL_Image is always downloaded from libsdl.org on Windows. 

How does this work with library dependencies? Will the travis.sh script run through in MSYS2? Because, this is how I create releases. 

> That miniz is not that mini, it feels weird to add so much irrelevant code to Doom port git.

We actually only need miniz.c and miniz.h from their release tarball (and the license file of course).

**Comment by rfomin** [comment]
> How does this work with library dependencies? Will the travis.sh script run through in MSYS2? Because, this is how I create releases.

It works fine from my testing.

> We actually only need miniz.c and miniz.h from their release tarball (and the license file of course).

Yes, but that miniz.c is over 200 kb. There is amalgamation.sh script but it looks like another build system :smile:

**Comment by fabiangreffrath** [comment]
> It works fine from my testing.

And does the game start from that directory in explorer? Pardon my asking, but from my experience, libraries in MSYS2 often have additional dependencies on some libgcc.dll and stuff like that. 

> Yes, but that miniz.c is over 200 kb. 

Still less than SDL_Image or libpng+zlib. 😛

**Comment by rfomin** [comment]
> And does the game start from that directory in explorer? Pardon my asking, but from my experience, libraries in MSYS2 often have additional dependencies on some libgcc.dll and stuff like that.

Yes, I tested that. Here is my MSYS2 package: [Woof-8.1.0-win64.zip](https://github.com/fabiangreffrath/woof/files/7739457/Woof-8.1.0-win64.zip)

**Comment by rfomin** [comment]
> And does the game start from that directory in explorer? Pardon my asking, but from my experience, libraries in MSYS2 often have additional dependencies on some libgcc.dll and stuff like that.

Actually, it only works on my dev machine. Sorry, I guess this is wrong approach.

**Comment by fabiangreffrath** [comment]
The approach *can* work, but we will need some semiautomatic library dependency tracking like this  
https://github.com/chocolate-doom/chocolate-doom/blob/master/pkg/win32/cp-with-libs

**Comment by rfomin** [comment]
> The approach _can_ work, but we will need some semiautomatic library dependency tracking like this https://github.com/chocolate-doom/chocolate-doom/blob/master/pkg/win32/cp-with-libs

I tried this - it turned out that only one dependency in libFLAC (libssp-0.dll) is required to make SDL_mixer work. There are some cmake solutions for this: [stackoverflow](https://stackoverflow.com/questions/62884439/how-to-use-cmake-file-get-runtime-dependencies-in-an-install-statement), but I can't get it to work. Oh well, I think this is a minor feature and not worth it probably.

File: Source/CMakeLists.txt [hunk 1]
```diff
@@ -190,7 +190,15 @@ if(WIN32)
...
     list(APPEND WOOF_DLLS "${SDL2_MIXER_DLL_DIR}/libFLAC-8.dll")
+    if(MINGW)
+       list(APPEND WOOF_DLLS "${SDL2_MIXER_DLL_DIR}/libssp-0.dll")
+    endif()
     list(APPEND WOOF_DLLS "${SDL2_MIXER_DLL_DIR}/libmodplug-1.dll")
+    if(MINGW)
+       list(APPEND WOOF_DLLS "${SDL2_MIXER_DLL_DIR}/libstdc++-6.dll")
+       list(APPEND WOOF_DLLS "${SDL2_MIXER_DLL_DIR}/libgcc_s_seh-1.dll")
+       list(APPEND WOOF_DLLS "${SDL2_MIXER_DLL_DIR}/libwinpthread-1.dll")
+    endif()
```
Comment (rfomin) [inline Source/CMakeLists.txt:201]: This is cheating but it works for me.

**Comment by fabiangreffrath** [comment]
Yes, this is cheating. 😛 This requires prior explicit knowledge of all dependencies - which is a moving target, especially on MSYS2. I like the idea to use cp-with-deps if Python is available, and keep the current solution of downloading SDL* and explicitly installing its bundled DLLs only for systems where this is required - i.e. mostly only MSVC.

File: Source/CMakeLists.txt [hunk 2]
```diff
@@ -217,3 +222,34 @@ else()
...
 endif()
 install(FILES ${WOOF_DLLS} DESTINATION .)
+
+if(MINGW)
+    install(CODE "set(CMAKE_MINGW_SYSTEM_LIBRARY_PATH \"${SDL2_MIXER_DLL_DIR}\")")
+
+    install(CODE [[
+        file(GET_RUNTIME_DEPENDENCIES
+            LIBRARIES "${CMAKE_MINGW_SYSTEM_LIBRARY_PATH}/SDL2_mixer.dll"
+                      "${CMAKE_MINGW_SYSTEM_LIBRARY_PATH}/libmodplug*.dll"
```
Comment (rfomin) [inline Source/CMakeLists.txt:232]: We can just get the dependencies for woof.exe, but then we need to get rid of the SDL_Image. Also, we need to add libmodplug anyway.

**Comment by rfomin** [comment]
Also, I have a Python script for this, but it doesn't look much better:
```cmake
install(CODE [[
    execute_process(
        COMMAND "${Python3_EXECUTABLE}" "${PROJECT_SOURCE_DIR}/mingw-bundledlls"
        "$<TARGET_FILE_DIR:woof>/woof.exe"
        OUTPUT_VARIABLE RUNTIME_DEPS
        OUTPUT_STRIP_TRAILING_WHITESPACE
        COMMAND_ECHO STDERR
        COMMAND_ERROR_IS_FATAL ANY
    )

    set(INSTALL_DIRS "${CMAKE_INSTALL_PREFIX};$<TARGET_FILE_DIR:woof>")
    foreach(INSTALL_DIR ${INSTALL_DIRS})
        file(INSTALL
            DESTINATION ${INSTALL_DIR}
            TYPE SHARED_LIBRARY
            FOLLOW_SYMLINK_CHAIN
            FILES ${RUNTIME_DEPS}
        )
    endforeach()
]])
```
cmake 3.14 is required for this.

**Comment by fabiangreffrath** [comment]
We would have to iterate through woof.exe's dependencies and install their dependencies likewise. This is what the cp-with-deps script from the Choco repo does.

**Comment by rfomin** [comment]
> We would have to iterate through woof.exe's dependencies and install their dependencies likewise. This is what the cp-with-deps script from the Choco repo does.

That is what GET_RUNTIME_DEPENDENCIES from cmake does. Choco/Crispy doesn't ship libmodplug.

**Comment by fabiangreffrath** [comment]
libmodplug is a special pony anyway, because it gets linked to SDL_Mixer at runtime, i.e. it isn't a hard depdendency but a "plugin".

**Comment by rfomin** [comment]
> libmodplug is a special pony anyway, because it gets linked to SDL_Mixer at runtime, i.e. it isn't a hard depdendency but a "plugin".

But we needed and it has some dependencies on MSYS2. The current solution works for me, but I haven't tested cross-compiling on Linux.

**Comment by fabiangreffrath** [comment]
> libmodplug is a special pony anyway, because it gets linked to SDL_Mixer at runtime, i.e. it isn't a hard depdendency but a "plugin".

Same as libfluidsynth, btw, which probably leads to the next step in this devilish plan. Can we define "soft"/optional dependencies for GET_RUNTIME_DEPENDENCIES?

**Comment by rfomin** [comment]
libfluidsynth has too many dependencies on MSYS2. Maybe we should download the official release from github. It works for me.

**Comment by fabiangreffrath** [comment]
Hm, but then we would still have to download some external resources, even on MSYS. ☹️

**Comment by rfomin** [comment]
> Hm, but then we would still have to download some external resources, even on MSYS. frowning_face

But it's good. I like the downloads from libsdl.org if they would update SDL_Mixer.

**Comment by rfomin** [comment]
It's actually not all that different. A total of 29 DLLs with fluidsynth from MSYS2 versus 28 DLLs from the official release. Our current release has 14 DLLs.

**Comment by fabiangreffrath** [comment]
3 less if we merge miniz branch. 😆

**Comment by rfomin** [comment]
> Does this include the useless libjpg, libtiff, libwebp, etc. that MSYS2's SDL_Image pulls in?

No, SDL_Image from libsdl.org. Most of the new DLLs come from libfluisynth and libmodplug.

**Comment by rfomin** [comment]
Full list:
```
libflac-8.dll        libmpg123-0.dll    libsndfile-1.dll     SDL2.dll
libfluidsynth-3.dll  libogg-0.dll       libssp-0.dll         SDL2_image.dll
libgcc_s_seh-1.dll   libopus-0.dll      libstdc++-6.dll      SDL2_mixer.dll
libglib-2.0-0.dll    libopusfile-0.dll  libtermcap-0.dll     SDL2_net.dll
libgomp-1.dll        libpcre-1.dll      libvorbis-0.dll      zlib1.dll
libiconv-2.dll       libpng16-16.dll    libvorbisenc-2.dll
libintl-8.dll        libportaudio.dll   libvorbisfile-3.dll
libmodplug-1.dll     libreadline8.dll   libwinpthread-1.dll
```

**Comment by fabiangreffrath** [comment]
So, is this now able to create a full install package for a Woof! release with all the EXE's dependencies and their respective dependencies in an MSYS2 environment without downloading additional external resources? Why does SDL2_Mixer need a special mention?

**Comment by rfomin** [comment]
> So, is this now able to create a full install package for a Woof! release with all the EXE's dependencies and their respective dependencies in an MSYS2 environment without downloading additional external resources? 

I did this specifically for SDL_Mixer, other SDL libraries are still downloaded from libsdl.org. I thought downloading was a feature, not a problem. If desired, we can easily disable downloading for MSYS2. Are we merging the miniz branch and are we including libfluidsynth?

**Comment by fabiangreffrath** [comment]
> I thought downloading was a feature, not a problem. 

It is an oddity that is strictly only required for MSVC builds (or for unprepared MSYS2 environments, but that's not our target). For Linux and MSYS2/MinGW, I'd prefer if we could bundle libraries from the actual build environment, because they carry patches that we may need, e. g. the fluidsynth crash patch for SDL2_Mixer.

> If desired, we can easily disable downloading for MSYS2. 

Yes, I think we should do this. We should only download any external resources if they cannot be found in the build environment.

> Are we merging the miniz branch 

I think so, yes.

> and are we including libfluidsynth?

Only as an optional dependency. If it is available, we should include it, but packaging should not fail if it is unavailable - a warning should suffice instead.

**Comment by rfomin** [comment]
Got it, I'll apply all needed changes after the miniz merge.

BTW, I think it would be nice to include fluidsynth with TimGM6mb.sf2 in the Woof distribution. Also, I want to add fluidsynth to the in-game menu.

**Comment by fabiangreffrath** [comment]
But isn't fluidsynth the default MIDI backend for SDL_Mixer anyway if a soundfont path is supplied? If we ship SDL_Mixer and fluidsynth and TimGM6mb.sf2 in the Woof distribution and point `fluidsynth_sf_path` to TimGM6mb.sf2, then SDL:Mixer should play all MIDI music with fluidsynth and this soundfont. Oh, wait, this collides with i_winmusic, right?

**Comment by rfomin** [comment]
> Oh, wait, this collides with i_winmusic, right?

Right. And for Windows users TimGM6mb.sf2 doesn't sound right. We've been listening MS GS Wavetable for 20+ years.

File: Source/CMakeLists.txt [hunk 3]
```diff
@@ -209,3 +209,66 @@ else()
     install(TARGETS woof RUNTIME DESTINATION bin)
 endif()
 install(FILES ${WOOF_DLLS} DESTINATION .)
+
+if(MINGW)
+    find_program(DLLTOOL_EXECUTABLE
```
Comment (fabiangreffrath) [inline Source/CMakeLists.txt:214]: Should even be safe for cross-compiling. Debian's mingw-w64 package contains a dlltool implementation:
https://packages.debian.org/search?searchon=contents&keywords=dlltool&mode=path&suite=stable&arch=any

**Review by fabiangreffrath — APPROVED** [review]

File: Source/CMakeLists.txt (hunk 3 above)
Comment (rfomin) [reply Source/CMakeLists.txt:214]: Oh sure. I haven't tested cross-compiling, but all this effort was mainly for this.

**Pull request merged**